app/charts/column/columns-simple.tsx

```diff
@@ -41,31 +43,92 @@ export const ErrorWhiskers = () => {
   );
 };
 
+type ChartDatum = {
+  x: number;
+  y: number;
+  width: number;
+  height: number;
+  color: string;
+};
+
 export const Columns = () => {
+  const ref = React.useRef<SVGGElement>(null);
   const { preparedData, bounds, getX, xScale, getY, yScale } =
     useChartState() as ColumnsState;
   const theme = useTheme();
   const { margins } = bounds;
 
-  return (
-    <g transform={`translate(${margins.left} ${margins.top})`}>
-      {preparedData.map((d, i) => {
-        const y = getY(d) ?? NaN;
-        const x = xScale(getX(d)) as number;
-        return (
-          <Column
-            key={i}
-            data-index={i}
-            width={xScale.bandwidth()}
-            x={x}
-            y={yScale(Math.max(y, 0))}
-            height={Math.abs(yScale(y) - yScale(0))}
-            color={
-              y <= 0 ? theme.palette.secondary.main : theme.palette.primary.main
-            }
-          />
+  const chartData: ChartDatum[] = React.useMemo(() => {
+    const bandwidth = xScale.bandwidth();
+    const getColor = (d: number) => {
+      return d <= 0 ? theme.palette.secondary.main : theme.palette.primary.main;
+    };
+
+    return preparedData.map((d) => {
+      const xScaled = xScale(getX(d)) as number;
+      const y = getY(d) ?? NaN;
+      const yScaled = yScale(y);
+      const height = Math.abs(yScaled - yScale(0));
+      const color = getColor(y);
+
+      return {
+        x: xScaled,
+        y: yScaled,
+        width: bandwidth,
+        height,
+        color,
+      };
+    });
+  }, [
+    preparedData,
+    getX,
+    getY,
+    xScale,
+    yScale,
+    theme.palette.primary.main,
+    theme.palette.secondary.main,
+  ]);
+
+  React.useEffect(() => {
+    if (ref.current) {
+      const g = select(ref.current);
+      g.selectAll<SVGRectElement, ChartDatum>("rect")
+        .data(chartData, (d) => d.x)
+        .join(
+          (enter) =>
+            enter
+              .append("rect")
+              .attr("data-index", (_, i) => i)
```

Comment:
What is the purpose of this data-index attribute ? I am not seeing it used in the update/exit transitions.

Reply:
Ah ok I see it is used in a test.

Reply:
👍